source/distributed/tria.cc

```diff
@@ -3122,6 +3122,204 @@ namespace parallel
 
 
 
+    namespace
+    {
+      template <typename ITERATOR>
+      void
+      identify_periodic_vertices_recursively(const GridTools::PeriodicFacePair<ITERATOR> &periodic,
+                                             std::vector<unsigned int> &topological_vertex_numbering)
+      {
+        // for hanging nodes we will consider all necessary coupling already
```

Comment:
Case 3 should be fine because on adaptive meshes there is no corresponding other vertex id of hanging nodes. The vertices which appear on both sides are identified on the next coarser level.